library/CMakeLists.txt

```diff
@@ -112,6 +112,39 @@ set(src_tls
     ssl_tls13_keys.c
 )
 
+find_package(Perl REQUIRED)
+
+file(GLOB error_headers ${CMAKE_CURRENT_SOURCE_DIR}/../include/mbedtls/*.h)
+add_custom_command(
+    OUTPUT
+        ${CMAKE_CURRENT_BINARY_DIR}/error.c
```

Comment:
(Response to Note 1) The normal CMake convention is to never touch the source tree at all, isn't it? So I think it does make sense to generate the source files in the build tree.

However, I think there should be two distinct ways of building the source files: by actually building them, or by copying `${CMAKE_CURRENT_SOURCE_DIR}/…` to `${CMAKE_CURRENT_BINARY_DIR}/…` (only applicable when these are not the same). Ideally copying should be done only if the files in the source tree are up-to-date, because copying out-of-date files would be surprising behavior.

Reply:
Yes, I agree, I was thinking about that over the week-end. I think we could use `link_to_source()` in the second build mode.

One thing that's not clear to me though is whether the distinction between the two build modes should be determined explicitly (for example, users of the "build generated files" mode would need to set some variable) or implicitly, based on whether the file exists in the source tree. I think explicit would be cleaner and easier here.

Reply:
I've been thinking about dependencies in the makefile too. Consider a user who downloads a release and makes a copy of the files without preserving timestamps. Then, depending on the exact order in which the files were copied, some generated files may be older than their source and the build system would try to regenerate them. This would be a problem if the user doesn't have all the required tools (which include not only perl and python, but also a _host_ C compiler, which the user might not have if they're cross-compiling the library).

So like you I'm beginning to think that the build scripts should be different, or have some different setting (e.g. based on the presence of a file) in release mode vs development mode. The rules to regenerate source files, or at least the dependencies, should only be active in development mode.

Reply:
Ok, I've handled the first part: we're no longer writing anything to the source tree.

I'll handle the second part (having a distinct mode for when the files shouldn't be generated) in a future round.

Reply:
I think I've handled the second part: distinguish between release mode and dev mode.